### PR TITLE
feat: [PLG-2486]: Changed to support for only one credit card

### PR DESCRIPTION
**Pull request by thomas-grasman**
## Describe your changes
Removed support for multiple credit cards and added functionality that limits accounts to one credit card at a time.

- Added a new endpoint that retrieves the primary credit card rather than a list (Will only be one for now)
- When we save a new card, we must first delete all other cards saved to the customer's stripe account.
- Added a fingerprint to credit card DTO. Fingerprints are hashed identifiers created by Stripe which will allow us to detect duplicate cards entered by multiple accounts.
- Added logging to the credit card service

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/49254)

### CODE REVIEW

**Comment by thomas-grasman** [comment]
trigger uts

**Comment by thomas-grasman** [comment]
trigger uts

**Comment by thomas-grasman** [comment]
trigger uts

**Review by Xin-Harness — DISMISSED** [review]

File: 930-ng-license-manager/src/main/java/io/harness/subscription/resource/CreditCardResource.java [hunk 1]
```diff
@@ -77,6 +81,22 @@ public ResponseDTO<CreditCardResponse> saveCreditCard(
     return ResponseDTO.newResponse(creditCardService.saveCreditCard(creditCardDTO));
   }
 
+  @GET
+  @Path("/primary-card")
```
Comment (neelam-harness) [inline 930-ng-license-manager/src/main/java/io/harness/subscription/resource/CreditCardResource.java:85]: Can we not make this api more generic and get hold of the primary card information by providing an additional query param ?
For eg:
generic endpoint could be, `/card` and primary card information could be fetched as `/card?type=primary` This way this API can be used for one or more use-cases.

**Review by thomas-grasman — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49254)* status: 8 file(s) to review. 1 change requests. 4. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`930-ng-license-manager/src/main/java/io/harness/subscription/resource/CreditCardResource.java` line 85 at r3](https://reviewable.io/reviews/harness/harness-core/49254#gh-1235630440:-NYTLO0V0yO8wl6bdche:b-8xyslh) ([raw file](https://github.com/harness/harness-core/blob/0f9358da3b7c06ae8c8226df857018972b5c9b3e/930-ng-license-manager/src/main/java/io/harness/subscription/resource/CreditCardResource.java#L85)):*
<details><summary><i>Previously, neelam-harness wrote…</i></summary><blockquote>

Can we not make this api more generic and get hold of the primary card information by providing an additional query param ?
For eg:
generic endpoint could be, `/card` and primary card information could be fetched as `/card?type=primary` This way this API can be used for one or more use-cases. 
</blockquote></details>

We already have an endpoint that retrieves a list of cards with one labeled as primary. This endpoint was only created because we want to support only one card for now and don't want to select from a list. So, the "primary" card will be the only card for now.
___
*[`930-ng-license-manager/src/main/java/io/harness/subscription/resource/CreditCardResource.java` line 97 at r3](https://reviewable.io/reviews/harness/harness-core/49254#-NYOcCut4wLBPtj-2MCh:-NYTUjaZ2COekTSFpeEC:bfjhbmf) ([raw file](https://github.com/harness/harness-core/blob/0f9358da3b7c06ae8c8226df857018972b5c9b3e/930-ng-license-manager/src/main/java/io/harness/subscription/resource/CreditCardResource.java#L97)):*
<details><summary><i>Previously, kapil-harness (Kapil Choudhary) wrote…</i></summary><blockquote>

either rename this method getPrimaryCreditCard or change the endpoint to "/default-credit-card"
</blockquote></details>

API endpoint changed to /credit-cards/default
___
*[`930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/CreditCardServiceImpl.java` line 59 at r3](https://reviewable.io/reviews/harness/harness-core/49254#-NYOeDkm6KzECTDEO99r:-NYTcXE98lS08UO6es3A:b-rmmhi8) ([raw file](https://github.com/harness/harness-core/blob/0f9358da3b7c06ae8c8226df857018972b5c9b3e/930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/CreditCardServiceImpl.java#L59)):*
<details><summary><i>Previously, kapil-harness (Kapil Choudhary) wrote…</i></summary><blockquote>

why are we deleting an existing card if creditCardRequest does not have appropriate accoundID.
Deleting the card is done below in the code, which is expected:

</blockquote></details>

If the accountId doesn't match the current customer's accountId, that means a different customer already saved this card to our system. When we see that's happened, we do not save it and we must remove it from Stripe's system as well.
___
*[`930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/CreditCardServiceImpl.java` line 115 at r3](https://reviewable.io/reviews/harness/harness-core/49254#-NYOglGL0GYpC7LJTozU:-NYTcTAuCvNPDJMy2Vd7:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/0f9358da3b7c06ae8c8226df857018972b5c9b3e/930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/CreditCardServiceImpl.java#L115)):*
<details><summary><i>Previously, kapil-harness (Kapil Choudhary) wrote…</i></summary><blockquote>

use this Lombok annotation to get rid of this incorrect naming getIsDefaultCard() or setIsDefaultCard.
Post this annotation you will be directly able to use isDefaultCard() for both getter and setter.

</blockquote></details>

Done.

File: 930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/SubscriptionServiceImpl.java [hunk 2]
```diff
@@ -657,11 +657,11 @@ private boolean checkSubscriptionInValid(SubscriptionDetail subscriptionDetail,
   }
 
   private void isSelfServiceEnable() {
-    if (deployMode.equals("KUBERNETES_ONPREM")) {
-      String errorMessage = "Self Service is not available for OnPrem deployments.";
-      log.error(errorMessage);
-      throw new UnsupportedOperationException(errorMessage);
-    }
+    //    if (deployMode.equals("KUBERNETES_ONPREM")) {
```
Comment (neelam-harness) [inline 930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/SubscriptionServiceImpl.java]: might have to remove the comments here

**Review by neelam-harness — APPROVED** [review]

**Comment by thomas-grasman** [comment]
trigger uts

**Review by neelam-harness — APPROVED** [review]

**Comment by thomas-grasman** [comment]
trigger uts

**Review by thomas-grasman — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49254)* status: 12 file(s) to review. 1 change requests. 4. unresolved discussions. 1 of 2 approvals obtained. 

___
*[`930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/CreditCardServiceImpl.java` line 115 at r3](https://reviewable.io/reviews/harness/harness-core/49254#-NYOglGL0GYpC7LJTozU:-NYV88MYE2frKJhYzPcS:b-z3ac5g) ([raw file](https://github.com/harness/harness-core/blob/0f9358da3b7c06ae8c8226df857018972b5c9b3e/930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/CreditCardServiceImpl.java#L115)):*
<details><summary><i>Previously, thomas-grasman (Thomas Grasman) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Using this accessor changed the return value from the object I needed to just one string from that object. I reverted this change for now. We can discuss this in more detail if you would like.

**Comment by thomas-grasman** [comment]
trigger uts

**Comment by thomas-grasman** [comment]
trigger uts

**Review by Xin-Harness — APPROVED** [review]

**Review by kapil-harness — APPROVED** [review]
> Reviewed 2 of 7 files at r1, 1 of 2 files at r2, 1 of 5 files at r4, 2 of 3 files at r5, 5 of 6 files at r6, 1 of 3 files at r7, 2 of 2 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49254)* status: 14 file(s) to review. 1. unresolved discussions. 

___
*[`930-ng-license-manager/src/main/java/io/harness/subscription/resource/CreditCardResource.java` line 97 at r3](https://reviewable.io/reviews/harness/harness-core/49254#-NYOcCut4wLBPtj-2MCh:-NZPe4E-0A6BZIW-DXD9:bu8517) ([raw file](https://github.com/harness/harness-core/blob/0f9358da3b7c06ae8c8226df857018972b5c9b3e/930-ng-license-manager/src/main/java/io/harness/subscription/resource/CreditCardResource.java#L97)):*
<details><summary><i>Previously, thomas-grasman (Thomas Grasman) wrote…</i></summary><blockquote>

API endpoint changed to /credit-cards/default
</blockquote></details>

I was trying to say that either use the term primary or default, not both, as it may be confusing.
___
*[`930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/CreditCardServiceImpl.java` line 59 at r3](https://reviewable.io/reviews/harness/harness-core/49254#-NYOeDkm6KzECTDEO99r:-NZPexXGDHZhRkYV21_w:b-ld7l7k) ([raw file](https://github.com/harness/harness-core/blob/0f9358da3b7c06ae8c8226df857018972b5c9b3e/930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/CreditCardServiceImpl.java#L59)):*
<details><summary><i>Previously, thomas-grasman (Thomas Grasman) wrote…</i></summary><blockquote>

If the accountId doesn't match the current customer's accountId, that means a different customer already saved this card to our system. When we see that's happened, we do not save it and we must remove it from Stripe's system as well.
</blockquote></details>

this seems like vey intrinsic to stripe, I may need to read a bit more on it
___
*[`930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/CreditCardServiceImpl.java` line 115 at r3](https://reviewable.io/reviews/harness/harness-core/49254#-NYOglGL0GYpC7LJTozU:-NZPfBRS6Z3QTv28-ZWE:b-b2um2x) ([raw file](https://github.com/harness/harness-core/blob/0f9358da3b7c06ae8c8226df857018972b5c9b3e/930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/CreditCardServiceImpl.java#L115)):*
<details><summary><i>Previously, thomas-grasman (Thomas Grasman) wrote…</i></summary><blockquote>

Using this accessor changed the return value from the object I needed to just one string from that object. I reverted this change for now. We can discuss this in more detail if you would like.
</blockquote></details>

strange, it should not change the return value at all. Lets connect on this.

**Pull request merged**